content/300-guides/100-performance-and-optimization/150-connection-management/index.mdx

```diff
@@ -0,0 +1,286 @@
+
+### <inlinecode>PrismaClient</inlinecode> in serverless environments
+
+#### Instantiate `PrismaClient` outside the handler
+
+Instantiate `PrismaClient` [outside the scope of the function handler](https://github.com/prisma/e2e-tests/blob/5d1041d3f19245d3d237d959eca94d1d796e3a52/platforms/serverless-lambda/index.ts#L3) to increase the chances of reuse. As long as the handler remains 'warm' (in use), the connection is potentially reusable:
+
+```ts highlight=3;normal
+import { PrismaClient, Prisma } from '@prisma/client'
+
+const client = new PrismaClient()
+
+export async function handler() {
+    /* ... */
+}
+```
+
+#### Do not explicitly <inlinecode>$disconnect()</inlinecode>
+
+You [do not need to explicitly `$disconnect()`](../../../concepts/components/prisma-client/working-with-prismaclient/connection-management#calling-disconnect-explicitly) <span class="concept"></span> at the end of a function, as there is a possibility that the container might be reused. Opening a new connection takes time and slows down your function's ability to process requests.
+
+### Other serverless considerations
+
+#### Container reuse
+
+There is no guarnatee that subsequent nearby invocations of a function will hit the same container - for example, AWS can choose to create a new container at any time.
+
+Code should assume the container to be stateless and create a connection only if it does not exist - Prisma Client JS already implements this logic.
+
+#### Zombie connections
+
+Containers that are marked to be removed and are not being reused still keep a connection open and can stay in that state for some time (unknown and not documented from AWS). This can lead to sub-optimal utilization of the DB connections.
+
+Potential solutions<sup>1, 2</sup> include:
+
+* Using a lower idle connection timeout.
+* Cleaning up the idle connections in a separate service.
+
+<sup>
+  1. Note that these are recommendations and not best practices. These would vary from system to
+  system.
+</sup>
+<br />
+<sup>
+  2.{' '}
+  <a href="https://github.com/jeremydaly/serverless-mysql">
+    <inlineCode>serverless-mysql</inlineCode>
+  </a>{' '}
+  is a library that implements this idea.
+</sup>
+
+#### Concurrency limits
+
+Depending on your serverless concurrency limit (the number of serverless functions running in parallel), you might still exhaust your database's connection limit. This can happen when too many functions are invoked concurrently, each with its own connection pool, which eventually exhausts the database connection limit. To prevent this, you can [set your serverless concurrency limit](https://docs.aws.amazon.com/lambda/latest/dg/configuration-concurrency.html) to a number lower than the connection limit of your database (as you might want to be able to connect from another client for other purposes).
```

Comment:
```suggestion
Depending on your serverless concurrency limit (the number of serverless functions running in parallel), you might still exhaust your database's connection limit. This can happen when too many functions are invoked concurrently, each with its own connection pool, which eventually exhausts the database connection limit. To prevent this, you can [set your serverless concurrency limit](https://docs.aws.amazon.com/lambda/latest/dg/configuration-concurrency.html) to a number lower than the maximum connection limit of your database divided by the connection used by each function invocation (as you might want to be able to connect from another client for other purposes).
```
bad wording here, but not everyone should be using pool size 1